model/project_ref.go

```diff
@@ -247,22 +250,67 @@ func (p *ProjectRef) Add(creator *user.DBUser) error {
 	return p.AddPermissions(creator)
 }
 
+func (p *ProjectRef) GetRepoId() string {
+	return fmt.Sprintf("%s_%s", p.Owner, p.Repo)
+}
+
+func (p *ProjectRef) AddToRepoScope(ctx context.Context, user *user.DBUser) error {
+	rm := evergreen.GetEnvironment().RoleManager()
+	repoId := p.GetRepoId()
+	repoRef, err := FindOneRepoRef(repoId)
+	if err != nil {
+		return errors.Wrapf(err, "error finding repo ref '%s'", repoId)
+	}
+	if repoRef == nil {
+		repoRef := RepoRef{ProjectRef{
+			Id:      repoId,
+			Admins:  []string{user.Username()},
+			Owner:   p.Owner,
+			Repo:    p.Repo,
+			Enabled: true,
+		}}
+		// creates scope and give user admin access to repo
+		return errors.Wrapf(repoRef.Add(user), "problem adding new repo ref '%s'", repoId)
+	} else {
+		// if the repo exists, then the scope also exists, so add this project ID to the scope, and give the user repo admin access
+		repoRole := GetRepoRole(repoId)
+		if !utility.StringSliceContains(user.Roles(), repoRole) {
+			if err = user.AddRole(repoRole); err != nil {
+				return errors.Wrapf(err, "error adding admin role to repo '%s'", user.Username())
+			}
+			repoRef.Admins = append(repoRef.Admins, user.Username())
+			if err = repoRef.Update(); err != nil {
+				return errors.Wrapf(err, "error adding user as repo admin")
+			}
+		}
+	}
+
+	return errors.Wrapf(rm.AddResourceToScope(GetRepoScope(repoId), p.Id), "error adding resource to repo '%s' scope", repoId)
```

Comment:
This is giving a naming convention more meaning than just as an identifier. Could you either look up the repo scope beforehand, or add something to gimlet that finds a scope with a single resource?

Reply:
But this is the scope's _id, not just an identifier. If I can't look up the scopes (and the roles for that matter) using the ID I know it has, then the find function you suggest I write will return the repo scope and the unrestricted scope, but I already know the ID so why do we have to do this? For the constants UnrestrictedProjects and AllProjects we're allowed to do a direct lookup; I feel like the inability to use the _id adds a layer of complication we don't necessarily need.

Reply:
If we're not okay with using _id for roles either (276) then we'd also have to make a new FindRoleWithPermissions that returns all roles that have the repo resource.